File: phoenix/etc/dewaterize.py
#!/usr/bin/env python

import sys
import os

functions = ( 
    'w_assert',
    '$water->Notice',
    '$water->Trace',
    '$water->Warning'
 )

def decommentized( f ):
    

def dewaterized( f ):
    new = []
    removeGlobal = True
    toAppend = True
    for line in f:
        stripped = line.strip()
        for func in functions:
            if stripped.startswith( func ):
                toAppend = False
        if toAppend:
            new.append( line )
        if stripped.endswith( ';' ):
            toAppend = True
    for line in new:
        if '$water' in line:
            removeGlobal = False
            break
    contents = removeGlobal and [ line for line in new if 'global $water' not in line ] or new
    return ''.join( contents )

def optimized( directory, extensions ):
    for root, subdirs, files in os.walk( directory ):
        try:
            files.remove( 'water.php' )
        except ValueError:
            pass
        for filename in files:
            for ext in extensions:
                if filename.endswith( ext ):
                    name = os.path.join( root, filename ) 
                    f = open( name )
                    filtered = dewaterized( decommentized( f.read() ) )
                    f.close()
                    f = open( name, 'w' )
                    f.write( filtered )
                    f.close()
                    break

if __name__ == '__main__':
    try:
        target = sys.argv[ 1 ]
    except IndexError:
        target = '.'
    exts = sys.argv[ 2: ] or ( 'php', )
    dewaterize( target, exts )

File: phoenix/etc/dewaterize.php
#!/usr/bin/env php

<?php

	function optimized( $source ) {
		$methods = array( 'Notice', 'Trace', 'Warning' );
		$new = '';
		$append = true;
		$tokens = token_get_all( $source );
		for ( $i = 0; $i < count( $tokens ); ++$i ) {
			if ( is_array( $tokens[ $i ] ) ) {
				$tokens[ $i ] = $tokens[ $i ][ 1 ];
			}
		}
		foreach ( $tokens as $i => $token ) {
			if ( $token == 'w_assert' ) {
				$append = false;
			}
			else if ( $token == '$water' && $tokens[ $i + 1 ] == '->' ) {
				foreach ( $methods as $meth ) {
					if ( $tokens[ $i + 2 ] == $meth ) {
						$append = false;
					}
				}
			}
			if ( $append ) {
				$new .= $token;
			}
			if ( $token == ';' ) {
				$append = true;
			}
		}
		return $new;
	}

	function dewaterize( $directory, $extensions ) {
		$files = New RecursiveIteratorIterator( New RecursiveDirectoryIterator( $directory ) );
		foreach ( $files as $file ) {
			$filename = $file->getFilename();
			if ( $filename != 'water.php' ) {
				foreach ( $extensions as $ext ) {
					if ( substr( $filename, strlen( $filename ) - strlen( $ext ) ) == $ext ) {
						file_put_contents( $file, optimized( file_get_contents( $file ) ) );
						break;
					}
				}
			}
		}
	}

	switch ( $argc ) {
		case 1:
			$directory = '.';
			$extensions = array( 'php' );
			break;
		case 2:
			$directory = $argv[ 1 ];
			$extensions = array( 'php' );
			break;
		default:
			$directory = $argv[ 1 ];
			$extensions = array_slice( $argv, 2 );
	}
	dewaterize( $directory, $extensions );

?>
